branch (#13999)
pingcap/docs: tidb-cloud/branch-overview.md

---
title: Branch overview
summary: Learn what is TiDB Serverless Branch.
---

> TiDB Serverless allows users to create branches of clusters. By using TiDB Serverless Branch, teams can work in parallel, iterate quickly on new features, troubleshoot issues without impacting the production database, and roll back changes easily if necessary. It aids in the overall development and deployment process while ensuring a high level of stability and reliability for the production database.

Reviewer: ```suggestion
TiDB Serverless lets you create branches of clusters. With TiDB Serverless Branch, teams can work in parallel, iterate rapidly on new features, troubleshoot issues without affecting the production database, and easily revert changes if needed. This feature streamlines the development and deployment process while ensuring a high level of stability and reliability for the production database.
```

## What is a Branch?

> A branch refers to a separate instance that contains a diverged copy of data from the cluster. It serves as an isolated environment, allowing users to freely experiment without any concerns about impacting the parent cluster. 

Reviewer: ```suggestion
A branch is a separate instance that contains a diverged copy of data from the cluster. It provides an isolated environment, allowing you to experiment freely without worrying about affecting the original cluster.
```

> At the point of creation, the data in the branch diverges from the cluster, meaning that subsequent changes made in either the cluster or the branch won't be synchronized. 

Reviewer: ```suggestion
When a branch is created, the data in it diverges from the original cluster, meaning that subsequent changes made in either the original cluster or the branch will not be synchronized.
```

> To ensure quick and seamless branch creation, TiDB Serverless utilizes a copy-on-write technique for sharing data between the cluster and branches. This process usually completes within several minutes and remains imperceptible to users, not affecting performance for the parent cluster.

Reviewer: ```suggestion
To ensure fast and seamless branch creation, TiDB Serverless uses a copy-on-write technique for sharing data between the original cluster and its branches. This process usually completes within a few minutes and remains imperceptible to users, ensuring that it does not affect the performance of the original cluster.
```

## Scenarios

> Branches are easy and quick to create and provide isolated data environments. It is beneficial in scenarios where multiple developers or teams need to work independently, test changes, fix bugs, experiment with new features, or roll out updates without disrupting the main production database.

Reviewer: ```suggestion
You can create branches easily and quickly to get isolated data environments. Branches are beneficial in the following scenarios where multiple developers or teams need to work independently, test changes, fix bugs, experiment with new features, or roll out updates without disrupting the main production database.
```

> 1. Feature Development: It enables developers to work on new features in isolation without affecting the main production database. Each feature can have its own branch, allowing developers to iterate quickly and experiment without impacting other ongoing work.

Reviewer: ```suggestion
- Feature Development: Developers can work on new features in isolation without affecting the main production database. Each feature can have its own branch, allowing quick iteration and experimentation without affecting other ongoing work.
```

> 2. Bug Fixing: Database branching aids in isolating and fixing bugs without introducing new issues to the main database. Developers can create a branch dedicated to fixing a specific bug, test the fix, and then merge it back once verified.

Reviewer: ```suggestion
- Bug Fixing: Developers can create a branch dedicated to fixing a specific bug, test the fix, and then merge it back once verified, without introducing new issues to the main database. 
```

> 3. Experimentation: While developing new features or making changes, developers can create branches to experiment with different approaches or configurations. This allows them to compare various options, gather data, and make informed decisions before merging changes into the main database.

Reviewer: ```suggestion
- Experimentation: While developing new features or making changes, developers can create branches to experiment with different approaches or configurations. This allows them to compare various options, gather data, and make informed decisions before merging changes into the main database.
```

> 4. Performance Optimization: Sometimes, database changes are made to enhance performance. With branching, developers can experiment and fine-tune different configurations, indexes, or algorithms in isolated environments to identify the most efficient solution.

Reviewer: ```suggestion
4. Performance optimization: Database changes are sometimes made to enhance performance. With branching, developers can experiment and fine-tune various configurations, indexes, or algorithms in isolated environments to identify the most efficient solution.
```

> 5. Testing and Staging: Database branching enables teams to create branches for testing and staging purposes. It ensures a controlled environment for quality assurance, user acceptance testing, or staging customizations before merging them into the main database.

Reviewer: ```suggestion
- Testing and staging: Teams can create branches for testing and staging purposes. It ensures a controlled environment for quality assurance, user acceptance testing, or staging customizations before merging them into the main database.
```

> 6. Parallel Development: Database branching allows different teams or developers to work on different projects simultaneously. Each project can have its own branch, allowing independent development and experimentation, while still being able to merge changes back into the main database.

Reviewer: ```suggestion
- Parallel development: Different teams or developers can work on separate projects simultaneously. Each project can have its own branch, enabling independent development and experimentation, while still being able to merge changes back into the main database.
```

> ## Limitations and Quotas

Reviewer: ```suggestion
## Limitations and quotas
```

> TiDB Serverless Branch is in Beta version. Currently, it is free of charge. 

Reviewer: ```suggestion
Currently, the TiDB Serverless branch feature is in beta and free of charge. 
```

> You can only create branches for TiDB Serverless clusters that are created after July 4, 2023.

Reviewer 2: ```suggestion
You can only create branches for TiDB Serverless clusters that are created after July 5, 2023.
```

> You can only create branches under clusters that are created after July 4, 2023.

Reviewer: ```suggestion
You can only create branches for TiDB Serverless clusters that are created after July 4, 2023.
```

---

> title: TiDB Serverless Branch Overview

Reviewer: ```suggestion
title: TiDB Serverless Branch (Beta) Overview
```

summary: Learn the concept of TiDB Serverless branches.
---

> # TiDB Serverless Branch Overview

Reviewer: ```suggestion
# TiDB Serverless Branch (Beta) Overview
```

> # TiDB Serverless Branch (Beta) Overview

Author: ```suggestion
# TiDB Serverless Branching (Beta) Overview
```

TiDB Cloud lets you create branches for TiDB Serverless clusters. A branch for a cluster is a separate instance that contains a diverged copy of data from the original cluster. It provides an isolated environment, allowing you to experiment freely without worrying about affecting the original cluster.

With TiDB Serverless branches, developers can work in parallel, iterate rapidly on new features, troubleshoot issues without affecting the production database, and easily revert changes if needed. This feature streamlines the development and deployment process while ensuring a high level of stability and reliability for the production database.

## Implementations

> When a branch for a cluster is created, the data in it diverges from the original cluster, meaning that subsequent changes made in either the original cluster or the branch will not be synchronized with each other.

Reviewer 3: ```suggestion
When a branch for a cluster is created, the data in the branch diverges from the original cluster. This means that subsequent changes made in either the original cluster or the branch will not be synchronized with each other.
```

> To ensure fast and seamless branch creation, TiDB Serverless uses a copy-on-write technique for sharing data between the original cluster and its branches. This process usually completes within a few minutes and remains imperceptible to users, ensuring that it does not affect the performance of your original cluster.

Reviewer 3: ```suggestion
To ensure fast and seamless branch creation, TiDB Serverless uses a copy-on-write technique for sharing data between the original cluster and its branches. This process usually completes within a few minutes and is imperceptible to users, ensuring that it does not affect the performance of your original cluster.
```

- Feature development: Developers can work on new features in isolation without affecting the production database. Each feature can have its own branch, allowing quick iteration and experimentation without affecting other ongoing work.

- Bug fixing: Developers can create a branch dedicated to fixing a specific bug, test the fix, and then merge it back once verified, without introducing new issues to the production database.

- Experimentation: While developing new features or making changes, developers can create branches to experiment with different approaches or configurations. This allows them to compare various options, gather data, and make informed decisions before the changes are merged into the production database.

- Performance optimization: Database changes are sometimes made to enhance performance. With branching, developers can experiment and fine-tune various configurations, indexes, or algorithms in isolated environments to identify the most efficient solution.

> - Testing and staging: Teams can create branches for testing and staging purposes. It ensures a controlled environment for quality assurance, user acceptance testing, or staging customizations before the changes are merged into the production database.

Reviewer: ```suggestion
- Testing and staging: Teams can create branches for testing and staging purposes. It provides a controlled environment for quality assurance, user acceptance testing, or staging customizations before the changes are merged into the production database.
```

- Parallel development: Different teams or developers can work on separate projects simultaneously. Each project can have its own branch, enabling independent development and experimentation, while still being able to merge changes back into the production database.

## Limitations and quotas

Currently, TiDB Serverless branches are in beta and free of charge.

- You can only create branches for TiDB Serverless clusters that are created after July 5, 2023.

> - For each organization in TiDB Cloud, you can create a maximum of five TiDB Serverless branches by default across all the clusters. The branches of a cluster will be created in the same region as the cluster, and you cannot create branches for a throttled cluster.

Reviewer 2: ```suggestion
- For each organization in TiDB Cloud, you can create a maximum of five TiDB Serverless branches by default across all the clusters. The branches of a cluster will be created in the same region as the cluster, and you cannot create branches for a throttled cluster or a cluster larger than 100 GiB.
```

Reviewer 2: @qiancai PTAL

Reviewer: "a cluster larger than 100 GiB" 是指行存+列存和到一起超过 100 GiB 吗

Author: 应该是指 row storage

Reviewer 2: row storage

- For each branch, 5 GiB storage is allowed. Once the storage is reached, the read and write operations on this branch will be throttled until you reduce the storage.

> If you need more quotas, [contact TiDB Cloud Support](/tidb-cloud/tidb-cloud-support.md).

Reviewer: "TiDB Cloud Support" is now used as a term for the support service in TiDB Cloud docs.
![image](https://github.com/pingcap/docs/assets/79440533/27173d25-3f95-4851-92a7-18caf13c798a)
